Bearded.Utilities.Tests/Collections/DeletableObjectListTests.cs

```diff
@@ -473,6 +473,106 @@ public void WorksNonGenerically()
                 enumeratedItems.Should().BeEquivalentTo(items, withExactSameItems);
             }
         }
+        
+        public class ReverseEnumeration
+        {
+            [Property]
+            public void SkipsDeletedItems(NonEmptyArray<bool> someBools)
+            {
+                var boolQueue = someBools.Get.Looping();
+                var (list, items) = createPopulatedList(20);
+
+                foreach (var item in list)
+                {
+                    if (boolQueue.Next())
+                        item.Deleted = true;
+                }
+
+                list.Reversed.Should().BeEquivalentTo(items.Reverse().Where(i => !i.Deleted), withExactSameItems);
+            }
+
+            [Fact]
+            public void SkipsItemsDeletedAheadOfEnumeration()
+            {
+                var (list, items) = createPopulatedList(20);
+                var enumeratedItems = new List<TestDeletable>();
+
+                foreach (var item in list.Reversed)
+                {
+                    enumeratedItems.Add(item);
+                    list.First().Deleted = true;
+                }
+
+                enumeratedItems.Should().BeEquivalentTo(items.Reverse().Take(10), withExactSameItems);
+            }
```

Comment:
This test's correctness relies quite a lot on getting logic and control structures right. It took me quite a long time to actually understand why this test is correct. That scares me a bit. The guarantee of correctness of the entire implementation relies on the tests being correct as well. That is straightforward if you use some simple examples, but as soon as you start using any control structures, suddenly I need to spend several minutes going through this method line by line to make sure it indeed behaves as we expect.

This is the case for most test methods added here. Some of them purposefully try to be a bit smart, but I don't think that adds to the value of the tests. In this particular case for example, I don't really see why the following test provides any less coverage:

```cs
[Fact]
public void SkipsItemsDeletedAheadOfEnumeration()
{
  // Arrange
  var (list, items) = createPopulatedList(20);
  list.First().Deleted = true;
 
  // Act
  var enumeratedItems = list.Reversed.ToList();

  // Assert
  enumeratedItems.Should().BeEquivalentTo(items.Reverse().Take(19), withExactSameItems);
}
```

If you specifically want to test that it skips enumerated items after the enumerator has been created, I would also make that more explicit:

```cs
[Fact]
public void SkipsItemsDeletedAheadOfEnumeration()
{
  // Arrange
  var (list, items) = createPopulatedList(20);
  var enumerator = list.Reversed.GetEnumerator();
  var enumeratedItems = new List<TestDeletable> { enumerator.Current };
  enumerator.MoveNext(); // make sure the enumerator has started doing things
  list.First().Deleted = true;
 
  // Act
  do
  {
    enumeratedItems.Add(enumerator.Current);
  } while (enumerator.MoveNext());

  // Assert
  enumeratedItems.Should().BeEquivalentTo(items.Reverse().Take(19), withExactSameItems);
}
```

Note that I also marked the three blocks of tests (arrange-act-assert) explicitly. See for example [this documentation](https://docs.microsoft.com/en-us/visualstudio/test/unit-test-basics?view=vs-2019#write-your-tests) on unit testing basics explaining this AAA structure, or [this blog](https://defragdev.com/blog/?p=783) explaining some of the benefits (tl;dr, it helps identifying code smells much easier).

Even these examples I gave already contain more logic and control structures than I'm comfortable with in unit tests (especially the second), but at least their use is minimised, making a test much more easier to verify on correctness, without sacrificing any test coverage in this particular case.